### ISSUE TITLE
Allow/support multiple `@JsonCreator` in single class

### ISSUE THREAD

**Issue by stasmihailov** [FasterXML/jackson-databind#2353]
I have an idea which I'd like to implement in this module. **BUT**, I want to hear other Jackson users' opinions on this idea first

---

Imagine I have a following class:

```java
class SuperToken {
    long time;
    String username;
    byte[] password;
}
```
 
So I'd like to deserialize it via dynamic deserializer selection:

### What i want to do:
```java

class SuperToken {
    long time;
    String username;
    byte[] password;
    
    @JsonCreator(mode=DELEGATING) // invoked when a string is passed
    public static SuperToken from(String username) {
        SuperToken token = new SuperToken();
        token.username = username;
        token.password = new byte[];
        token.time = System.currentTimeMillis();
        
        return token;
    }
    
    @JsonCreator(mode=PROPERTIES) // invoked when an object is passed, pre-validating property existence
    public static SuperToken create(
            String username,
            byte[] password,
            long time) {
        SuperToken token = new SuperToken();
        token.username = username;
        token.password = new byte[];
        token.time = time;

        return token;
    }
}
```

Unfortunately, this is not possible right now (as explained 
[once on Stackoverflow](https://stackoverflow.com/questions/46793301/multiple-jsoncreator-annotated-methods?noredirect=1&lq=1))
and [then in your Github issues](https://github.com/fasterxml/jackson-databind/issues/421))

### What i HAVE to do (now):
```java

class SuperToken {
    long time;
    String username;
    byte[] password;
    
    @JsonCreator(mode=PROPERTIES)
    public static SuperToken create(
            JsonNode node) {
        if (node.isTextual()) {
            String username = node.asText();
            return ...
        } else {
            // all this <strong>BORING</strong> deserialization
        }
    }
}
```

**Comment by kirillgroshkov** [comment]
We are also searching for a way to achieve that

**Comment by cowtowncoder** [comment]
@stasmihailov @kirillgroshkov It should already work, actually, as long as delegating case is scalar-valued (`int`/`long`/`String`/`boolean`). I added a test in 2.10 branch to verify its functioning.
The only change to test case is to add `@JsonProperty` for parameters of properties-case (or use Java 8 parameter names module).

**Comment by stasmihailov** [comment]
@kirillgroshkov indeed, if I have such code:
```java
@EqualsAndHashCode
public class CoolClass {
    private final String someString;
    private final int someInt;
    private final Map<String, String> someFields;

    CoolClass(String someString, int someInt, Map<String, String> someFields) {
        this.someString = someString;
        this.someInt = someInt;
        this.someFields = someFields;
    }

    @JsonCreator(mode = DELEGATING)
    public static CoolClass fromString(String someString) {
        return new CoolClass(someString, 0, emptyMap());
    }

    @JsonCreator(mode = PROPERTIES)
    public static CoolClass create(
            @JsonProperty("myString") String someString,
            @JsonProperty("myInt") int someInt,
            @JsonProperty("myFields") Map<String, String> someFields) {
        return new CoolClass(someString, someInt, someFields);
    }
}
```
these tests pass:
```java
class CoolClassTest {
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    static Arguments[] args() {
        return new Arguments[]{
                Arguments.of(
                        "\"str\"",
                        new CoolClass("str", 0, Collections.emptyMap())),
                Arguments.of(
                        // language=json
                        "{\"myString\": \"something\", \"myInt\": 14, \"myFields\": {}}",
                        new CoolClass("something", 14, Collections.emptyMap())),
        };
    }

    @ParameterizedTest
    @MethodSource("args")
    void create(String jsonSource, CoolClass expectedResult) throws Exception {
        CoolClass actualResult = OBJECT_MAPPER.readValue(jsonSource, CoolClass.class);
        assertEquals(expectedResult, actualResult);
    }
}
```
tested on 2.10.1

**Comment by cowtowncoder** [comment]
@stackmagic thank you for verifying this.

**Comment by Legion2** [comment]
When the single argument of the delegation JsonCreator is of Type Map<String, ...>, the Property JsonCreator is never used.

**Comment by cowtowncoder** [comment]
@Legion2 if you have a new problem, you want to create a new issue. We do not usually re-open closed issued.
Delegating creators can definitely use `Map`s too (but so can properties based ones).

**Comment by joffrey-bion** [comment]
I'm hitting this problem, and I just realized it's not possible at the moment to define multiple @JsonCreators in general. Shouldn't this issue stay open? I mean, it only works in a very specific case with only one property-based creator and one delegating creator that happens to take a scalar type. I guess the issue should either be renamed or be kept open.

I'm trying to add overloaded constructors for backwards compatibility with older JSONs and I'm at a loss at the moment. I tried to use multiple property-based creators with distinct property sets (for different old versions and the new version of the JSON), and it doesn't seem to work (Jackson complains that it already has other creators and doesn't want to add more). Any ideas how I can achieve that?
